src/lib/x509/pkix_enums.h

```diff
@@ -100,18 +102,92 @@ BOTAN_PUBLIC_API(2,0) const char* to_string(Certificate_Status_Code code);
-   DECIPHER_ONLY      = 1 << 7
-};
+class BOTAN_PUBLIC_API(3,0) Key_Constraints
+   {
+   public:
+      enum Bits : uint32_t {
+         None              = 0,
+         DigitalSignature  = 1 << 15,
+         NonRepudiation    = 1 << 14,
+         KeyEncipherment   = 1 << 13,
+         DataEncipherment  = 1 << 12,
+         KeyAgreement      = 1 << 11,
+         KeyCertSign       = 1 << 10,
+         CrlSign           = 1 << 9,
+         EncipherOnly      = 1 << 8,
+         DecipherOnly      = 1 << 7,
+
+         // Deprecated SHOUTING_CASE names for Key_Constraints
+         // will be removed in a future major release
+         NO_CONSTRAINTS    BOTAN_DEPRECATED("Use None")             = None,
+         DIGITAL_SIGNATURE BOTAN_DEPRECATED("Use DigitalSignature") = DigitalSignature,
+         NON_REPUDIATION   BOTAN_DEPRECATED("Use NonRepudiation")   = NonRepudiation,
+         KEY_ENCIPHERMENT  BOTAN_DEPRECATED("Use KeyEncipherment")  = KeyEncipherment,
+         DATA_ENCIPHERMENT BOTAN_DEPRECATED("Use DataEncipherment") = DataEncipherment,
+         KEY_AGREEMENT     BOTAN_DEPRECATED("Use KeyAgreement")     = KeyAgreement,
+         KEY_CERT_SIGN     BOTAN_DEPRECATED("Use KeyCertSign")      = KeyCertSign,
+         CRL_SIGN          BOTAN_DEPRECATED("Use CrlSign")          = CrlSign,
+         ENCIPHER_ONLY     BOTAN_DEPRECATED("Use EncipherOnly")     = EncipherOnly,
+         DECIPHER_ONLY     BOTAN_DEPRECATED("Use DecipherOnly")     = DecipherOnly,
+      };
+
+      Key_Constraints(const Key_Constraints& other) = default;
+      Key_Constraints(Key_Constraints&& other) = default;
+      Key_Constraints& operator=(const Key_Constraints& other) = default;
+      Key_Constraints& operator=(Key_Constraints&& other) = default;
+
+      Key_Constraints(Key_Constraints::Bits bits) : m_value(bits) {}
+
+      explicit Key_Constraints(uint32_t bits) : m_value(bits) {}
+
+      Key_Constraints() : m_value(0) {}
+
+      /**
+      * Return typical constraints for a CA certificate, namely
+      * KeyCertSign and CrlSign
+      */
+      static Key_Constraints ca_constraints()
+         {
+         return Key_Constraints(Key_Constraints::KeyCertSign | Key_Constraints::CrlSign);
+         }
+
+      bool operator==(const Key_Constraints&) const = default;
+
+      void operator|=(Key_Constraints::Bits other)
+         {
+         m_value |= other;
+         }
+
+      // Return true if all bits in mask are set
+      bool includes(Key_Constraints::Bits other) const { return (m_value & other) == other; }
+      bool includes(Key_Constraints other) const { return (m_value & other.m_value) == other.m_value; }
+
+      // Return true if any of the bits provided are set
+      bool includes_any(auto&& ...bits) const
+         {
+         for(auto bit: std::initializer_list<Key_Constraints::Bits>{ bits... })
+            {
+            if(this->includes(bit))
+               return true;
+            }
+         return false;
+         }
```

Comment:
Nice, I didn't know about `std::initializer_list<>` like that. However, I think that's a perfect case for C++17 fold expressions:

```suggestion
      bool includes_any(auto&& ...bits) const
         {
         return (m_value & (bits | ...)) > 0;
         }
```